### PR TITLE
tests: userspace: fix flaky behavior

### PR DESCRIPTION
**Pull request by tyrantlizard77**
- No longer call ztest_test_pass() out of a fatal exception,
  as if this took place on some child thread, the next test
  case could start on another CPU before the child has exited,
  leading to issues if the child thread object is recycled

- Get rid of some unnecessary synchronization semaphores.
  Use the scheduler and/or k_thread_join() instead.

- Simplify tests for read/write other threads not to spawn
  a child thread and then take a fatal fault on the ztest
  thread

- Add set_fault() clear_fault() as I do not enjoy typing.
  Despite these variables being voliatile, a barrier is
  needed to prevent re-ordering around non-volatile memory
  access

- Don't call ztest_test_pass() from child thread in
  test_user_mode_enter() due to possible races

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Review by ceolin — APPROVED** [review]

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 1]
```diff
@@ -143,9 +137,8 @@ static void test_write_control(void)
 #elif defined(CONFIG_ARC)
 	unsigned int er_status;
 
-	expect_fault = true;
-	expected_reason = K_ERR_CPU_EXCEPTION;
-	BARRIER();
+	set_fault(K_ERR_CPU_EXCEPTION);;
```
Comment (jenmwms) [inline tests/kernel/mem_protect/userspace/src/main.c:140]: Please remove extra semi-colon on this line

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 2]
```diff
@@ -248,9 +236,8 @@ static void test_write_kernro(void)
 	zassert_true(ptr < _image_rodata_end &&
 		     ptr >= _image_rodata_start,
 		     "_k_neg_eagain is not in rodata");
-	expect_fault = true;
-	expected_reason = K_ERR_CPU_EXCEPTION;
-	BARRIER();
+	set_fault(K_ERR_CPU_EXCEPTION);;
```
Comment (jenmwms) [inline tests/kernel/mem_protect/userspace/src/main.c:239]: extra semi-colon here too

**Review by jenmwms — APPROVED** [review]
minor fixes, otherwise LGTM

**Comment by andrewboie** [comment]
@jenmwms fixed thanks

**Pull request merged**